oadp-operator/controllers/velero.go

```diff
@@ -1,14 +1,469 @@
+	GCP_SHARED_CREDENTIALS_FILE   = "GCP_SHARED_CREDENTIALS_FILE"
+	//TODO: Add Proxy env vars
+	HTTP_PROXY  = "HTTP_PROXY"
+	HTTPS_PROXY = "HTTPS_PROXY"
+	NO_PROXY    = "NO_PROXY"
+)
+
+//TODO: Add Image customization options
+// Images
+const (
+	VELERO_IMAGE           = "quay.io/konveyor/velero:latest"
+	OPENSHIFT_PLUGIN_IMAGE = "quay.io/konveyor/openshift-velero-plugins:latest"
+	AWS_PLUGIN_IMAGE       = "quay.io/konveyor/velero-plugin-for-aws:latest"
+	AZURE_PLUGIN_IMAGE     = "quay.io/konveyor/velero-plugin-for-microsoft-azure:latest"
+	GCP_PLUGIN_IMAGE       = "quay.io/konveyor/velero-plugin-for-gcp:latest"
+	CSI_PLUGIN_IMAGE       = "quay.io/konveyor/velero-plugin-for-csi:latest"
+)
+
+// Plugin names
+const (
+	VELERO_PLUGIN_FOR_AWS       = "velero-plugin-for-aws"
+	VELERO_PLUGIN_FOR_AZURE     = "velero-plugin-for-microsoft-azure"
+	VELERO_PLUGIN_FOR_GCP       = "velero-plugin-for-gcp"
+	VELERO_PLUGIN_FOR_CSI       = "velero-plugin-for-csi"
+	VELERO_PLUGIN_FOR_OPENSHIFT = "openshift-velero-plugin"
 )
 
 func (r *VeleroReconciler) ReconcileVeleroDeployment(log logr.Logger) (bool, error) {
 	velero := oadpv1alpha1.Velero{}
 	if err := r.Get(r.Context, r.NamespacedName, &velero); err != nil {
 		return false, err
 	}
+
+	// Build Velero Deployment
+	veleroDeployment := r.buildVeleroDeployment(&velero)
+
+	// Create or Update Velero Deployment
+	op, err := controllerutil.CreateOrUpdate(r.Context, r.Client, veleroDeployment, func() error {
+
+		// Set controller reference to Velero controller
+		err := controllerutil.SetControllerReference(&velero, veleroDeployment, r.Scheme)
+		if err != nil {
+			return err
+		}
+
+		// TODO: check for Velero deployment status condition errors and respond here
+		// TODO: Not sure about the mutation function here
+		veleroDeployment = r.buildVeleroDeployment(&velero)
+		return nil
+	})
+
+	if err != nil {
+		return false, err
+	}
+
+	if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated {
+		// Trigger event to indicate Velero Deployment was created or updated
+		r.EventRecorder.Event(veleroDeployment,
+			corev1.EventTypeNormal,
+			"VeleroDeploymentReconciled",
+			fmt.Sprintf("performed %s on velero deployment %s/%s", op, veleroDeployment.Namespace, veleroDeployment.Name),
+		)
+	}
 	return true, nil
 }
+
+// Build Velero Deployment
+func (r *VeleroReconciler) buildVeleroDeployment(velero *oadpv1alpha1.Velero) *appsv1.Deployment {
+
+	veleroVolumeMounts := r.getVeleroVolumeMounts(velero)
+	veleroVolumes := r.getVeleroVolumes(velero)
+	veleroEnv := r.getVeleroEnv(velero)
+	veleroInitContainers := r.getVeleroInitContainers(velero)
+	veleroResourceReqs := r.getVeleroResourceReqs(velero)
+
+	deployment := appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      Velero,
+			Namespace: VeleroNamespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			//TODO: add velero nodeselector, needs to be added to the Velero CR first
+			Selector: &metav1.LabelSelector{
```

Comment:
https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates

I think you will have to be careful here. You might want to accept the pointer to deployment here and change the parts that you need based on whether it's a create or update.

Reply:
+1 you could be deleting labels by doing this